### ISSUE TITLE
ProxyLayoutRenderer when meeting unknown layout

### ISSUE THREAD

**Issue by snakefoot** [NLog/NLog#3557]
Can see a lot of people who doing the following:

- Using NLog Config Variables before they are declared at runtime
- Register custom NLog LayoutRenderer after config has been loaded.

Right now the NLog-config loader discards unknown layout-renderers from the Layout.

My thought would be that it creates an instance of ProxyLayoutRenderer, that constantly tries to lookup a valid NLog Config Variable or NLog Layout Renderer for every logevent. When finally finding a valid object, then it is used from that point.

Could probably create some throttle logic, that it only attempts dynamic lookup for every logevent the first 10 sec. And then it performs dynamic lookup every 1 second.

**Comment by 304NotModified** [comment]
hey!

I was think by allowing "lazy late bound" renderers. So if not known then it's substituted by a "lazy contruct". Then when logging, we try to substitute the "lazy" with the actual renderer. (also throttling allowed)

**Comment by snakefoot** [comment]
Well that would complicate the SimpleLayout logic. Need a steady hand to replace layoutrenderer while multiple logger threads are active. But I guess lazy construct can notify SimpleLayout about recompile is needed. Then just flip the active array when new compile is ready.


Sent from my Sony Xperia


---- Julian Verdurmen wrote ----


hey!

I was think by allowing "lazy late bound" renderers. So if not known then it's substituted by a "lazy contruct". Then when logging, we try to substitute the "lazy" with the actual renderer. (also throttling allowed)

**Comment by 304NotModified** [comment]
> My thought would be that it creates an instance of ProxyLayoutRenderer, that constantly tries to lookup a valid NLog Config Variable or NLog Layout Renderer for every logevent. When finally finding a valid object, then it is used from that point.

Maybe this is also the same as I proposed. Not sure. First I read it as "check every X seconds", but it's "check every X logevents", isn't?

I would be nice if we could use this approach (in the future) also for targets and that is needed for DI. (as NLog is loaded before the DI container is configured)

**Comment by snakefoot** [comment]
Nlog can write 200000 logevents per second. And sometimes someone just write logevents once an hour. So I think it should check every 1 second after a 10 second period where it has checked for ever logevent.


Sent from my Sony Xperia


---- Julian Verdurmen wrote ----


My thought would be that it creates an instance of ProxyLayoutRenderer, that constantly tries to lookup a valid NLog Config Variable or NLog Layout Renderer for every logevent. When finally finding a valid object, then it is used from that point.

Maybe this is also the same as I proposed. Not sure. First I read it as "check every X seconds", but it's "check every X logevents", isn't?

I would be nice if we could use this approach (in the future) also for targets and that is needed for DI. (as NLog is loaded before the DI container is configured)

**Comment by 304NotModified** [comment]
another option is to make the reload easier (easier to find and easier to use)?

**Comment by snakefoot** [comment]
> another option is to make the reload easier (easier to find and easier to use)?

Config reload has a lot of side-effects. Especially if you also have places where you modify config at runtime.

**Comment by 304NotModified** [comment]
True , but polling seems non transparent (unclear when and maybe also unexpected)

**Comment by snakefoot** [comment]
Could probably also make an event one could listen for, when an NLog-Config-Variable is updated, or a custom NLog-Layout-Renderer is registered. Maybe the event could just be a timestamp on the NLog-config, and the lazy construct checks if the timestamp has changed on every logevent (then it is very cheap to check for every logevent).

**Comment by 304NotModified** [comment]
Sounds good. The timestamp check is only in non initialized renderers (the proxy) isn't?

**Comment by bambams** [comment]
Since this is 2 years old and doesn't appear to be being worked on:

See [1] for a simple workaround also. Duplicated here, you can reload the config after registering your custom types:

    LogManager.Configuration = LogManager.Configuration.Reload();
    LogManager.ReconfigExistingLoggers();

[1] https://stackoverflow.com/questions/70428704/custom-nlog-layout-renderer-produces-no-output-in-some-projects?noredirect=1#comment124516754_70428704

**Comment by snakefoot** [comment]
Yes also have a hard time figuring out how to make any of the "solutions" work together with `ThrowConfigExceptions = true`.
